src/main.cpp

```diff
@@ -2528,7 +2568,58 @@ bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockIndex* pin
     blockundo.vtxundo.reserve(block.vtx.size() - 1);
     int nChecked = 0;
     int nOrphansChecked = 0;
-    LOCK(cs_xval);
+    arith_uint256 nStartingChainWork = chainActive.Tip()->nChainWork;
+
+    // Create a vector for storing hashes that will be deleted from the unverified and perverified txn sets.
+    // We will delete these hashes only if and when this block is the one that is accepted saving us the unnecessary 
+    // repeated locking and unlocking of cs_xval.
+    std::vector<uint256> vHashesToDelete;
+
+    // Create a temporary view of the UTXO set
+    CCoinsViewCache viewTempCache(pcoinsTip);
+
+    // Section for boost scoped lock on the scriptcheck_mutex
+    boost::thread::id this_id(boost::this_thread::get_id());
+    {
+
+    // Get the next available mutex and then use it to find the associated scriptcheckqueue. Then lock this thread
+    // with the mutex so other thread can used this scriptcheckqueue until all inputs have been checked and the 
+    // scriptcheckqueue threads have all returned.
+    boost::shared_ptr<boost::mutex> scriptcheck_mutex = allScriptCheckQueues.GetScriptCheckMutex();
+    if (fParallel) cs_main.unlock(); // unlock cs_main, we may be waiting here for a while before aquiring the scoped lock below
+    boost::mutex::scoped_lock lock(*scriptcheck_mutex);
+
+    // Get the next available scriptcheckqueue.
+    CCheckQueue<CScriptCheck>* pScriptQueue = allScriptCheckQueues.GetScriptCheckQueue(scriptcheck_mutex);
```

Comment:
Alright it seems that this whole thing ends up being the major design problem. I believe this needs to be fixed before PV moves forward. You are juggling with locks here and mixing that with block validation. This is brittle.

This should be refactored. Here is the API I would aim for :
````
auto token = signatureChecker.check(transactions_to_check);

// Do various stuff while signature are being validated...

auto result = token.get(); // Wait for transactions to be validated and return the result of the validation.
token.cancel(); // Cancel processing.
````

The token is a value type. The destructor does cancel the work if it hasn't finished already.

Behind this API you can hide pretty much whatever is wanted. The important thing is that it is abstracted away.